tardis/visualization/widgets/custom_abundance.py

```diff
@@ -0,0 +1,1700 @@
+import math
+import numpy as np
+import pandas as pd
+import ipywidgets as ipw
+import plotly.graph_objects as go
+from astropy import units as u
+from pyne import nucname
+
+from tardis.util.base import quantity_linspace
+from tardis.io.config_reader import Configuration
+from tardis.model import Radial1DModel
+from tardis.model.density import (
+    calculate_power_law_density,
+    calculate_exponential_density,
+)
+from tardis.io.config_validator import validate_dict
+from tardis.io.parsers.csvy import load_csvy
+from tardis.io.model_reader import (
+    read_uniform_abundances,
+    parse_csv_abundances,
+)
+from tardis.util.base import atomic_number2element_symbol, quantity_linspace
+from tardis.visualization.tools.convergence_plot import transition_colors
+
+import asyncio
+
+
+class Timer:
+    """An object to implement debounce using an asynchronous loop.
+
+    Notes
+    -----
+    This class is reproduced from ipywidgets documentation, for more information
+    please see https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20Events.html
+    """
+
+    def __init__(self, timeout, callback):
+        """Initialize the Timer with delay time and delayed function.
+
+        Parameters
+        ----------
+            timeout : float
+            callback : function
+        """
+        self._timeout = timeout
+        self._callback = callback
+
+    async def _job(self):
+        await asyncio.sleep(self._timeout)
+        self._callback()
+
+    def start(self):
+        self._task = asyncio.ensure_future(self._job())
+
+    def cancel(self):
+        self._task.cancel()
+
+
+def debounce(wait):
+    """Decorator that will postpone a function's execution until after
+     `wait` seconds have elapsed since the last time it was invoked.
+
+    Parameters
+    ----------
+        wait : float
+
+    Returns
+    -------
+        function
+
+    Notes
+    -----
+    This decorator is reproduced from ipywidgets documentation, for more information
+    please see https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20Events.html
+    """
+
+    def decorator(fn):
+        timer = None
+
+        def debounced(*args, **kwargs):
+            nonlocal timer
+
+            def call_it():
+                fn(*args, **kwargs)
+
+            if timer is not None:
+                timer.cancel()
+            timer = Timer(wait, call_it)
+            timer.start()
+
+        return debounced
+
+    return decorator
+
+
+import yaml
+
+YAML_DELIMITER = "---"
+
+
+class CustomYAML(yaml.YAMLObject):
+    """A custom YAML object generated by required properties."""
+
+    def __init__(
+        self, name, d_time_0, i_time_0, v_inner_boundary, v_outer_boundary
+    ):
+        """Initialize CustomYAML object with given properties.
+
+        Parameters
+        ----------
+            name : str
+                Name of the YAML file.
+            d_time_0 : astropy.units.quantity.Quantity
+                Model density time 0.
```

Comment:
This is the initial time for the density in the model - see https://tardis-sn.github.io/tardis/io/configuration/components/models/index.html?highlight=csvy#csvy-model for the descriptions of this and i_time_0. It would be good to update both of these to have a more informative description